keras_hub/src/models/gemma3/gemma3_causal_lm_preprocessor.py

```diff
@@ -448,39 +535,69 @@ def generate_preprocess(
         the sequence (as generation is expected to continue at the end of the
         inputted prompt).
         """
+
         if not self.built:
             self.build(None)
 
+        # Extract inputs.
         if isinstance(x, dict):
             images = x.get("images", None)
-            num_valid_images = x.get("num_valid_images", None)
+
             # TODO: do we even need `responses` for generation? Makes sense for
-            # finetuning (i.e., `call()`).
+            # finetuning only (i.e., `call()`).
             responses = x.get("responses", None)
             prompts = x["prompts"]
         else:
             images = None
-            num_valid_images = None
             responses = None
             prompts = x
 
-        if self.text_only_model:
-            if images is not None or num_valid_images is not None:
-                raise ValueError(
-                    "`image_converter` cannot be None when `images` or"
-                    " `num_valid_images` is not None."
-                )
-        else:
-            # Replace `"<start_of_image>"` in prompts with
-            # `"\n\n<start_of_image> <img> * 256 <end_of_image>\n\n"`.
+        # Find out if the input is batched/not batched. Uprank if not batched.
+        # In other preprocessors, we don't have to do this, but here, all
+        # the following logic (indices, etc.) uses tensors with a batch dim.
+        # We will squeeze these back at the end.
+        batched = True
+        if isinstance(prompts, str):
+            batched = False
+            prompts = [prompts]
+            if responses is not None:
+                responses = [responses]
+        if isinstance(prompts, tf.Tensor) and len(prompts.shape) == 0:
+            batched = False
+            prompts = tf.expand_dims(prompts, axis=0)
+            if responses is not None:
+                responses = tf.expand_dims(responses, axis=0)
+
+        # There are 8 cases, based on values of
```

Comment:
Consider if there's some common utilities to refactor and share between call/generate. I see a lot of duplicated code.